scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -142,62 +125,264 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+        node->_less = _less;
+        node->less = buf.data() + node->_less;
+        node->_greater = _greater;
+        node->greater = buf.data() + node->_greater;
 
         return node_index;
     }
 }
 
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+    double *maxes, double *mins,
+    const int _median, const int _compact,
+    std::list<std::vector<ckdtreenode>>* buflist, int workers);
+
+inline ckdtree_intp_t
+build_recurse(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+    double *maxes, double *mins,
+    const int _median, const int _compact,
+    std::list<std::vector<ckdtreenode>>* buflist, int workers)
+{
+    if(workers > 1) {
+        return build(
+            self, start_idx, end_idx, maxes, mins, _median, _compact,
+            buflist, workers);
+    }
+    else {
+        return build(
+            self, start_idx, end_idx, maxes, mins, _median, _compact,
+            buflist->front());
+    }
+}
+
+struct build_worker {
+
+    std::list<std::vector<ckdtreenode>> buflist_;
+    std::vector<double> mins_, maxes_;
+    std::future<ckdtree_intp_t> future_;
+
+    ckdtree_intp_t run(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+          const double *maxes, const double *mins,
+          const int _median, const int _compact,
+          int workers)
+    {
+        const int subworkers = workers/2;
+        if(subworkers > 0) {
+            mins_.resize(self->m);
+            std::copy_n(mins, mins_.size(), mins_.begin());
+            maxes_.resize(self->m);
+            std::copy_n(maxes, maxes_.size(), maxes_.begin());
+
+            future_ = std::async(std::launch::async, [&]() {
+                    buflist_.emplace_back();
+                    buflist_.back().reserve((end_idx-start_idx)/self->leafsize/workers);
+                    return build_recurse(
+                        self, start_idx, end_idx, maxes_.data(), mins_.data(), _median, _compact,
+                        &buflist_, subworkers);
+                });
+        }
+        return subworkers;
+    }
+
+    operator bool() const { return future_.valid(); }
+    void wait() { future_.wait(); }
+    std::list<std::vector<ckdtreenode>>& buflist() { return buflist_; }
+};
+
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+      double *maxes, double *mins,
+      const int _median, const int _compact,
+      std::list<std::vector<ckdtreenode>>* buflist, int workers)
+{
+    std::vector<ckdtreenode>& buf = buflist->front();
+
+    /* put a new node into the node stack */
+    const ckdtree_intp_t node_index = buf.size();
+    ckdtree_intp_t _less;
+    buf.emplace_back();
+    ckdtreenode *node = &buf.back();
+    init_ckdtreenode(*node, start_idx, end_idx);
+
+    if (end_idx-start_idx <= self->leafsize) {
+        /* below brute force limit, return leafnode */
+        node->split_dim = -1;
+        return node_index;
+    }
```

Comment:
A heuristic will always have cases where it does not make the optimal choice. This decision is also only relevant for small trees which will for one usually be run with one worker and where the performance does not matter that much (the docstring of `KDTree` states somewhere, that most time is usually spend in querying). In the case of an extremely imbalanced tree, where this case could be hit on the short branch of a large tree, this also does not matter, because the runtime will be dominated by the worker processing the deep branch.

The user can control this behavior using the `workers` parameter, if they care. However, if someone is building many small trees, they are better off using `workers=1` and build the trees in parallel in any case.

This suggestion optimizes for an irrelevant case in an inflexible way, which is why I oppose it.

Reply:
>This suggestion optimizes for an irrelevant case in an inflexible way, which is why I oppose it.

It also makes the code simpler, so even if you don't care about efficiency it's still cleaner code.

Reply:
The early check of `workers` does make it simpler, but I prefer not to add the small subtree part.

Reply:
Added early `workers==1` check and call to sequential version of `build()` to eliminate `build_recurse()`.